Add ArcSinhTransformer, TextFeatures, and GeoDistanceTransformer

ankitlade12 opened the pull request:
## Summary

This PR adds three new transformers to Feature-engine, expanding the library's capabilities for numerical transformation, text feature extraction, and geospatial data processing.

---
## New Transformers
### 1. ArcSinhTransformer (transformation module)
**Addresses Issue #707**
Applies the inverse hyperbolic sine (arcsinh) transformation to numerical variables. Also known as the pseudo-logarithm, this transformation is useful for data that contains both positive and negative values.
**Key Features:**
- Handles zero and negative values (unlike LogTransformer)
- Supports [loc](cci:1://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_transformation/test_arcsinh.py:22:4-31:65) and [scale](cci:1://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_transformation/test_arcsinh.py:22:4-31:65) parameters for centering and normalization
- Full [inverse_transform](cci:1://file:///Users/ankithemantlade/Desktop/feature_engine/feature_engine/transformation/log.py:174:4-198:16) support
- For large |x|, behaves like ln(|x|) + ln(2), providing similar variance-stabilizing properties as log
**Use Cases:** Financial data (profit/loss), sensor readings, any metric that can be positive or negative
---
### 2. TextFeatures (NEW text module)
Extracts 19 numerical features from text/string columns. Unlike scikit-learn's CountVectorizer or TfidfVectorizer which create sparse matrices, TextFeatures extracts metadata features that remain in DataFrame format.
**Available Features:**
- `char_count`, `word_count`, [sentence_count](cci:1://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_text/__init__.py:93:4-99:64), `avg_word_length`
- `digit_count`, `uppercase_count`, `lowercase_count`, `special_char_count`
- `whitespace_count`, `whitespace_ratio`, `digit_ratio`, `uppercase_ratio`
- `has_digits`, `has_uppercase`, `is_empty`
- `starts_with_uppercase`, `ends_with_punctuation`
- `unique_word_count`, `unique_word_ratio`
**Use Cases:** NLP preprocessing, sentiment analysis feature engineering, text classification
---
### 3. GeoDistanceTransformer (creation module)
**Addresses Issue #688**
Calculates the distance between two geographical coordinate pairs (latitude/longitude) and adds the result as a new feature.
**Distance Methods:**
- **Haversine:** Great-circle distance (default, most accurate)
- **Euclidean:** Simple coordinate distance (fast)
- **Manhattan:** Taxicab distance (grid approximation)
**Output Units:** km, miles, meters, feet
**Validation:** Checks latitude range (-90 to 90) and longitude range (-180 to 180)
**Use Cases:** Real estate pricing, delivery optimization, ride-sharing, location-based ML
---
## What's Included
- Full test coverage with 43 passing tests
- RST documentation for all three transformers
- Code style compliance verified with flake8, black, and isort
- NumPy-style docstrings with usage examples
- All transformers follow Feature-engine's existing patterns and conventions
---
## Files Added/Modified
**New Files:**
- [feature_engine/transformation/arcsinh.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/feature_engine/transformation/arcsinh.py:0:0-0:0)
- [feature_engine/text/__init__.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/feature_engine/text/__init__.py:0:0-0:0)
- [feature_engine/text/text_features.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/feature_engine/text/text_features.py:0:0-0:0)
- [feature_engine/creation/geo_features.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/feature_engine/creation/geo_features.py:0:0-0:0)
- [tests/test_transformation/test_arcsinh.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_transformation/test_arcsinh.py:0:0-0:0)
- [tests/test_text/__init__.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_text/__init__.py:0:0-0:0)
- [tests/test_creation/test_geo_features.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_creation/test_geo_features.py:0:0-0:0)
- [docs/user_guide/transformation/ArcSinhTransformer.rst](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/docs/user_guide/transformation/ArcSinhTransformer.rst:0:0-0:0)
- [docs/user_guide/text/index.rst](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/docs/user_guide/text/index.rst:0:0-0:0)
- [docs/user_guide/text/TextFeatures.rst](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/docs/user_guide/text/TextFeatures.rst:0:0-0:0)
- [docs/user_guide/creation/GeoDistanceTransformer.rst](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/docs/user_guide/creation/GeoDistanceTransformer.rst:0:0-0:0)

**Modified Files:**
- [feature_engine/transformation/__init__.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/feature_engine/transformation/__init__.py:0:0-0:0)
- [feature_engine/creation/__init__.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/feature_engine/creation/__init__.py:0:0-0:0)
- [docs/user_guide/transformation/index.rst](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/docs/user_guide/transformation/index.rst:0:0-0:0)
- [docs/user_guide/creation/index.rst](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/docs/user_guide/creation/index.rst:0:0-0:0)
- [docs/user_guide/index.rst](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/docs/user_guide/index.rst:0:0-0:0)

## CI Status Note
The test failures in `test_feature_engine_py312` and `py313` are **pre-existing issues** in the main branch, not caused by this PR.
I verified this by:
1. Checking out the upstream `main` branch (without my changes)
2. Running `pytest tests/test_base_transformers/test_get_feature_names_out_mixin.py`
3. The same 73 tests fail there as well
All 27 tests for the new transformers pass:
- [test_arcsinh.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_transformation/test_arcsinh.py:0:0-0:0): 11 passed
- [test_geo_features.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_creation/test_geo_features.py:0:0-0:0): 16 passed  
- [test_text/__init__.py](cci:7://file:///Users/ankithemantlade/Desktop/feature_engine/tests/test_text/__init__.py:0:0-0:0): Pass
The `stylechecks`, `test_type`, and `test_docs` checks should all pass.

solegalli commented on feature_engine/creation/geo_features.py:
```diff
@@ -0,0 +1,437 @@
+    Methods
+    -------
+    fit:
+        This transformer does not learn parameters. Validates input columns.
```
```suggestion
        This transformer does not learn parameters.
```

solegalli commented on feature_engine/creation/geo_features.py:
```diff
@@ -0,0 +1,437 @@
+    get_feature_names_out:
+        Get output feature names for transformation.
+
+    See Also
```
I am not sure we need this section. These transformers are not related to GeoDistance's functionality.

solegalli commented on feature_engine/creation/geo_features.py:
```diff
@@ -0,0 +1,437 @@
+        """
+        This transformer does not learn parameters.
+
+        Validates that the coordinate columns exist and are numerical.
```
If I remember correctly, we don't normally include this sentence in our transformers. The library has grown quite a bit, so I might be wrong. I fail to keep track of everything :)

solegalli commented on tests/test_creation/test_geo_features.py:
```diff
@@ -0,0 +1,266 @@
+from feature_engine.creation import GeoDistanceTransformer
+
+
+class TestGeoDistanceTransformer:
```
What's the advantage of creating a class? 

We'd normally have the tests individually, ie, outside the class.

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+.. _geo_distance_transformer:
+
+.. currentmodule:: feature_engine.creation
```
We also need to add corresponding files to `docs/api/..` so that the documentation is created from the class docstrings.

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+GeoDistanceTransformer
+======================
+
+The :class:`GeoDistanceTransformer()` calculates the distance between two geographical
```
```suggestion
:class:`GeoDistanceTransformer()` calculates the distance between two geographical
```

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+and any domain where geographic proximity is relevant.
+
+Distance Methods
+~~~~~~~~~~~~~~~~
```
this should be - instead of ~

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+  Useful as a rough approximation for grid-based city layouts.
+
+Output Units
+~~~~~~~~~~~~
```
this should be - instead of ~

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+
+Using different distance methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Please add a sentence preceding the code explaining what will the code do.

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+    print(X_transformed.columns.tolist())
+    # ['trip_id', 'geo_distance']
+
+Using in a Pipeline
```
```suggestion
Calculating distance within a Pipeline
```

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+    # ['trip_id', 'geo_distance']
+
+Using in a Pipeline
+~~~~~~~~~~~~~~~~~~~
```
This will need adjusting to match the new heading length.

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+~~~~~~~~~~~~~~~~~~~
+
+:class:`GeoDistanceTransformer()` works seamlessly with scikit-learn pipelines:
+
```
please add a sentence indicating what's going on in the code.

solegalli commented on docs/user_guide/creation/GeoDistanceTransformer.rst:
```diff
@@ -0,0 +1,169 @@
+        )),
+        ('scaler', StandardScaler()),
+        ('regressor', RandomForestRegressor())
+    ])
```
please fit and show some output :)

ankitlade12 commented:
> Hey @ankitlade12
> 
> Thank you so much for these fantastic additions! They were much needed and long overdue :)
> 
> I only reviewed the files relevant for the GeoDistance transformer. The transformer is almost ready to be merged. I have a couple of nitpicks, a question regarding whether we should expose the optionality of a sanity check, and some more nitpicks on the documentation.
> 
> Would you be able to take a look at that?
> 
> In the ideal world, it would be great to have 1 PR per transformer. Any chance you could distribute the files relevant to each new transformer in different PRs?
> 
> If not, we take it from here. Thanks a lot!
> 
> In the meantime, I'll try and fix the tests in a separate PR. They are probably caused by the new version of sklearn :_(

ankitlade12 commented:
Hi @solegalli,

Thank you for the detailed review! I've addressed all the feedback for the GeoDistanceTransformer:

geo_features.py:

Added 
validate_ranges
 parameter (default True) to allow users to control coordinate validation
Updated docstrings to match library patterns (removed validation-related sentences)
test_geo_features.py:

Refactored tests from class-based to standalone functions
Added tests for the new 
validate_ranges
 parameter
User Guide (GeoDistanceTransformer.rst):

Fixed heading levels (- instead of ~ for main sections)
Changed "Example" to "Python Demo"
Updated class references per style guide
Added explanatory text before all code blocks
Split distance method examples with individual outputs
Renamed "Converting to miles" to "Using different output units"
Updated pipeline section with explanatory sentence, fit, and output
Removed API Reference section
API Documentation:

Created 
docs/api_doc/creation/GeoDistanceTransformer.rst
Added to 
docs/api_doc/creation/index.rst

Regarding separate PRs per transformer - I'll keep them in this single PR for now if that's okay. Let me know if there's anything else that needs adjustment!

solegalli commented on tests/test_creation/test_geo_features.py:
```diff
@@ -0,0 +1,306 @@
+    assert "geo_distance" in X_tr.columns
+
+
+def test_validate_ranges_parameter_validation():
```
here we'd add a pytest command to test more than one value, for Example "True" and a number or something else. 1 and 0 will pass though, because they are interpreted as booleans

solegalli commented on tests/test_creation/test_geo_features.py:
```diff
@@ -0,0 +1,306 @@
+
+    feature_names = transformer.get_feature_names_out()
+    assert "geo_distance" in feature_names
+    assert len(feature_names) == 6  # 5 original + 1 new
```
it would be better to test that the feature names are the expected ones.

ankitlade12 commented:
We have separated branches based on individual feature
